AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/DataStoreConfiguration.swift

```diff
@@ -0,0 +1,112 @@
+//
+// Copyright 2018-2020 Amazon.com,
+// Inc. or its affiliates. All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Amplify
+import Foundation
+
+/// Error Handler function typealias
+public typealias DataStoreErrorHandler = (AmplifyError) -> Void
+
+/// Holds a reference to both the local `Model` and the remote one during a conflict
+/// resolution. Implementations of the `DataStoreConflictHandler` use this to decide
+/// what the outcome of a conflict should be.
+public struct DataStoreConclictData {
```

Comment:
typo here; `DataStoreConflictData`

Reply:
I'm discovering there is a difference here from JS. API returns conflictUnhandled with error.data as the MutationSync<AnyModel> object. This means it gets deserialized into MutationSync containing the model and syncMetadata fields like `version` and `deleted`. In JS, this is returned as error.data with the `deleted` field on the model itself and passed to the confliict handler.

I'm guessing the developer can then decide what to do with the two models based on this information. The `DataStoreConflictData` doesn't pass the "is this model deleted?" information back.


Now that brings back to the question, what should DataStore do based on the result of the ConflictHandler?

If developer decides to `discard` the local changes, and the remote model is deleted, we should probably delete it from local store. On a side note, iif the remote model is not deleted, then we update local store with the remote model and update syncMetadata item in the local DB with the remote.version

If developer decides to `keep` the local changes, and the remote model is deleted, then we should call API.update(localModel, remote.version). the model is still marked as deleted after we update it. This could possibly be a bug, how do we possibly un-delete a model?

If the developer decides to give us a `new` model, then we should call API.update(newModel, remote.version), this is the same scenario as keeping the local changes. DataStore will call API and rely on the sync path to reconcile local changes.

if the conflict occurs at the service with the remote model deleted, should we still allow the client to resolve the conflict? If they decide to keep the local model, how do we "undo" a deleted remote model?

Reply:
thanks for pointing that out @lawmicha 
I'll merge this PR with that limitation, but feel free to update the `conflictHandler` signature to match your needs if you still need the extra data. After our conversation on Friday I was not 100% clear if we need the mutation sync field or just the models when resolving the conflicts.

Reply:
(I inadvertently deleted a comment here from @drochetti -- I've pinged him to reinstate it. Sorry for that!)